### ISSUE TITLE
Support `UNNEST` within joins, §7.6

### ISSUE THREAD

**Issue by jarohen** [xtdb/crux#2084]
Currently supported within `FROM`, but not (e.g.) `INNER JOIN UNNEST`.

* TBD: should we fix this, or just close it down for now (so that it gives a decent error message)

----
(from @mbutlerw)

Currently we plan out collection-derived-tables (UNNEST) in isolation and then re-arrange the various tables into the correct order (see build-table-reference-list). This works for simple cases, however I think we should opt for manual recursive planning, similar to how qualified joins are planned if we want to handle more complicated queries.

This is more speculative, but we currently use the `:unwind` operator in place of an explicit join. Perhaps it would be better to do an apply join (paraming the required cols if needed) with the newly unwound relation on the RHS, which should make UNNEST compatable with any join/query shape.

**Comment by mbutlerw** [comment]
Currently we plan out collection-derived-tables (UNNEST) in isolation and then re-arrange the various tables into the correct order (see build-table-reference-list). This works for simple cases, however I think we should opt for manual recursive planning, similar to how qualified joins are planned if we want to handle more complicated queries.

This is more speculative, but we currently use the `:unwind` operator in place of an explicit join. Perhaps it would be better to do an apply join (paraming the required cols if needed) with the newly unwound relation on the RHS, which should make UNNEST compatable with any join/query shape.

**Comment by jarohen** [comment]
re-check after #3383

**Comment by danmason** [comment]
Wrote up the following test:
```
(deftest test-unnest-in-join
  (xt/submit-tx tu/*node* [[:put-docs :actors
                            {:xt/id "bob", :name "Bob", :films ["The Great Escape", "Bob the Builder"]}]])
  
  (t/is (= [{:name "Bob", :film "The Great Escape"}
            {:name "Bob", :film "Bob the Builder"}]
           (xt/q tu/*node* "SELECT a.name, films.film 
                            FROM actors a 
                            JOIN UNNEST(a.films) AS films(film) on true"))))
```

Observing the following:
```
; 13:38:38.996 [nREPL-session-891bd36e-b29c-4514-b4d2-3dc5fba9dcfe] WARN  xtdb.sql.plan - Column not found: a.films
```

We then get an error later on:
```
; ERROR in xtdb.sql-test/test-unnest-in-join (ValueVectorReader.java:366):
; error: java.lang.UnsupportedOperationException: org.apache.arrow.vector.NullVector (ValueVectorReader.java:366)
; expected:
```

**Comment by refset** [comment]
Another example (I think?)

[Returns no results in XT](https://play.xtdb.com/?version=2.x-SNAPSHOT&type=sql-v2&enc=2&txs=NobwRAzgnhAuCmBbAtLAlo%2BYBcA7ArgDaEA0YsAHhDmAJIByAygKIBKAKgAQPsDynAEwD2AYwidWzAMK9WAEUacQAfTQDsnAIwlOIoYU0aQAIw0ByYwEMATmYC%2BdgNxg7JcAEd88a1BqxrXmBklNTYYCwAMtJclgB08IRIADq4AGKsvACygqIQOgCq9PTMjOwAFMJisXoGAJScAIKKlmUJSLXOruAhNADqtOwAEpwADtZCAvgisOJNnGXpWfMAag0R%2BSUpnPPanGYRliOwQiNmOg2srA0AmsCaAAw6AEyPnADM9wC6tSRb8086MwABQAFkJcPAzo1LjdgC8dAAWL4-P5lN6A9iWYyJWBQi5XW4PHQAVmRtT%2Bcwo81UAh0uEsmB0sEsAHMILVybhItFRrEAFQpRbZMYTKYzUYpXqDNjMTgvbj0ebcqRcZksmmcIWcQrFUplEaxNXsxqKChlNU0jnOMjQOBIVAYLB4IiEOyfIA):
```
INSERT INTO docs RECORDS {_id: 1, col1: {b: 'bar'}};
SELECT a.elem
FROM docs, UNNEST(docs.col1) AS a(elem);
```

But [works in Postgres](https://dbfiddle.uk/p-5zrX2v):
```
-- 1. Define a composite type ("struct")
CREATE TYPE my_struct AS (
  b text
);

-- 2. Create a table that uses an array of that struct
CREATE TABLE docs (
  id   int,
  col1 my_struct[]
);

-- 3. Insert a row using PostgreSQL record literal syntax
INSERT INTO docs VALUES (
  2,
  ARRAY[ROW('bar')::my_struct]
);

-- 4. Query unnesting the array of structs
SELECT a.elem
FROM docs, UNNEST(docs.col1) AS a(elem);
```

**Comment by refset** [comment]
Noting that you can workaround this (very inefficiently, most likely) by wrapping with LATERAL, e.g. ([Play](https://play.xtdb.com/?version=2.x-SNAPSHOT&type=sql-v2&enc=2&txs=NobwRAzgnhAuCmBbAtLAlo%2BYBcA7ArgDaEA0YsAHhDmAJIByAygKIBKAKgAQPsDynABwBOAewAm%2BAMawInVswDCvVgBFGnEAH00Y7JwCMJQULST4e-QDoADAF8A3AB1cDFh270%2Bx8VJlzFymoa2rqcAExGwqbm4TZGkiKEIkJ6AORC8GKpDs6ubFw8-MI%2B0rLySqrqWjp6AMyRJmZ1cZyIaBCSeiAC%2BABGhO0AFpmaAIaweuy0ALLMjOwAgtMACpypYdZhAKzI1vq7%2BgBa2Tm4uUz5Hl7JYvBCZQGVwTWcACxGA7jwaAiIEHrAQzhAC6DjAthI4Eo1GwYAAYqxeNNvBJSvYwGRoHAkKgMFg8ERCBCoVQaCwADKKLgAKhIzk4nHoc3YmmmC3oAE0ABQUqmcakM5gADQU5IAqipmJwBv56EspVziqiZJYQpwFuoBCEAJR03AMg2Go1M%2BaaXhMrkIpGcG53WQAdQAEmwFbb7pZPt9fhBtYDgZwALwo3wQVU6bX0o1Rw0am1CW73TQAd2GGU0St82ggmgAZmh7rBI9GDVbkeSFuw2AtyZxLYjkWL6Cb2Fy3aHPT8kD71eoBlyBtrtdK0EXiwaAFK8BjB0qcc21mVBjOlMNiQc94dfTt-SOluMJiDozEwX64zA4AjEWzAoA)):
```
SELECT *,
  NEST_MANY(SELECT *  EXCLUDE li RENAME (products._id AS p_id)
            FROM LATERAL (FROM UNNEST(orders.lineitems) AS li(li)) li
              JOIN products ON (li = products._id)) AS lineitems
  FROM orders;
```

Ideally it would be possible like:
```
SELECT *,
  NEST_MANY(SELECT *  EXCLUDE li RENAME (products._id AS p_id)
            FROM UNNEST(o.lineitems) AS li
              JOIN products AS p ON p._id = li
  ) AS lineitems_expanded
FROM orders AS o;
```

**Comment by jarohen** [comment]
`UNNEST` works in every join shape we'd expect it to. Closing this as done — with one piece of the original TBD spun out, since it's the half that's still open.

## tl;dr

- **The original complaint no longer reproduces.** `INNER JOIN UNNEST(…)` plans and executes, as does every other join type, in both the correlated and uncorrelated cases.
- **The mechanism this issue blames is gone.** There is no `build-table-reference-list` any more, and `visitCollectionDerivedTable` threads a `left-scope` — which is the recursive planning @mbutlerw proposed here, rather than plan-in-isolation-then-rearrange.
- **Two shapes still fail, and both should.** `LATERAL UNNEST(…)` isn't grammatical, and a correlated `UNNEST` can't be the right operand of a `RIGHT`/`FULL JOIN`.
- **The TBD's second half survives as #5872.** "Close it down so it gives a decent error message" is still owed for the `RIGHT`/`FULL` case, which leaks an internal symbol.

## What was verified

30 shapes, driven over XTDB's in-process ADBC connection against an in-memory node, split by whether the `UNNEST` argument correlates with the left-hand side.

**Uncorrelated — `UNNEST(?)` of a bound list.** All pass.

| Shape | Notes |
| --- | --- |
| `FROM t, UNNEST(?)` | comma join |
| `JOIN UNNEST(?) ON` | the shape this issue reports as broken |
| `INNER JOIN UNNEST(?) ON` | ditto, explicit `INNER` |
| `LEFT JOIN UNNEST(?) ON` | null-extends non-matching left rows |
| `RIGHT JOIN UNNEST(?) ON` | null-extends unmatched list elements |
| `FULL OUTER JOIN UNNEST(?) ON` | both sides null-extend |
| `CROSS JOIN UNNEST(?)` | |
| `FROM UNNEST(?) AS k(x) JOIN t ON` | `UNNEST` as the **left** operand |
| `JOIN UNNEST(?) WITH ORDINALITY ON` | |
| `JOIN UNNEST(?) ON` under `FOR VALID_TIME AS OF ? FOR SYSTEM_TIME AS OF ?` | |
| `JOIN UNNEST(?) AS k(r) ON t.a = (k.r).a AND t.b = (k.r).b` | list of structs, composite predicate |
| two `JOIN UNNEST(?)`s in one query | |

**Correlated — `UNNEST(t.tags)`.** Eight pass, five fail.

| Shape | |
| --- | --- |
| `FROM t, UNNEST(t.tags)` | passes |
| `JOIN UNNEST(t.tags) ON true` | passes |
| `INNER JOIN UNNEST(t.tags) ON k.x = 'red'` | passes — predicate on the unnested column |
| `LEFT JOIN UNNEST(t.tags) ON true` | passes — a row with `[]` null-extends rather than vanishing |
| `CROSS JOIN UNNEST(t.tags)` | passes |
| `NATURAL JOIN UNNEST(t.tags)` | passes |
| `JOIN UNNEST(t.tags) WITH ORDINALITY ON true` | passes |
| `UNNEST` inside a nested subquery that is then joined | passes |
| `JOIN LATERAL UNNEST(t.tags) ON true` | **parse error** |
| `LEFT JOIN LATERAL UNNEST(t.tags) ON true` | **parse error** |
| `RIGHT JOIN UNNEST(t.tags) ON true` | **leaks an internal symbol** |
| `FULL JOIN UNNEST(t.tags) ON true` | **leaks an internal symbol** |
| `FROM UNNEST(t.tags) AS k(x) JOIN t ON true` | correctly rejected: `Column not found: t.tags` |

## The failures, and why only one of them is a problem

**`LATERAL UNNEST(…)` is not grammatical, and needn't be.**

```
line 1:48 extraneous input 'UNNEST' expecting '('
```

`Sql.g4` has `'LATERAL' subquery`, and `subquery` is a parenthesised query expression — so `LATERAL` composes with `(SELECT …)` and not with `UNNEST`. That's the right reading of the standard, where a collection derived table is already implicitly lateral: `JOIN UNNEST(t.tags) AS k(x) ON true` is the supported spelling and it works. Postgres does accept `LATERAL` in front of a function call, so this is a compatibility nicety rather than a gap in expressiveness — worth an issue only if someone actually trips over it. The error is precise about position and token, so they'd find out quickly.

**Backward correlation is rejected correctly.** `FROM UNNEST(t.tags) AS k(x) JOIN prices AS t` gives `Column not found: t.tags`, which is what a lateral reference to a later `FROM` item should give.

**`RIGHT`/`FULL JOIN` over a correlated `UNNEST` is the one worth fixing.**

```
Unknown symbol: '?_sq_tags_1'
```

The rejection is correct — a correlated right operand has no rows independent of the left, so there is nothing to null-extend against — but `?_sq_tags_1` is the internal parameter the apply join threads in, and a user has no way to connect it to `t.tags`. Uncorrelated `RIGHT`/`FULL JOIN UNNEST(?)` is unaffected and works, so this is specifically about the correlated case.

That's exactly the "or just close it down for now (so that it gives a decent error message)" half of this issue's TBD, so it carries over as #5872 rather than being dropped.

## Coverage

Worth being straight about what is and isn't pinned by a test, since the matrix above came from a throwaway probe that has been deleted.

- **Already covered**: `unnest-ordinality-join-issue-4131` in `src/test/clojure/xtdb/operator/unnest_test.clj` asserts a correlated `JOIN UNNEST(r2.xs) WITH ORDINALITY … ON i1 = i2`, and `src/test/clojure/xtdb/sql_test.clj` has a further `JOIN UNNEST` case. So the headline claim of this issue is regression-tested.
- **Not covered**: the join-type matrix — `LEFT`/`RIGHT`/`FULL`/`CROSS`, `UNNEST` in left-operand position, and the bound-parameter `UNNEST(?)` variants.
- **Not the same thing**: the `food-prices` benchmark (#5858) leans on `UNNEST(?)` of a bound list heavily, but as `IN (SELECT k.x FROM UNNEST(?) AS k(x))` — a subquery, not a join operand — so it doesn't exercise these shapes.

Locking the matrix into `unnest_test.clj` is cheap if we want it; say the word and I'll raise it.
